common/src/main/cpp/extensions/openxr_fb_hand_tracking_aim_extension_wrapper.cpp

```diff
@@ -0,0 +1,164 @@
+	XRServer *xr_server = XRServer::get_singleton();
+
+	for (int i = 0; i < Hand::HAND_MAX; i++) {
+		if (xr_server != nullptr) {
+			xr_server->remove_tracker(trackers[i]);
+		}
+		trackers[i].unref();
+	}
+
+	fb_hand_tracking_aim_ext = false;
+}
+
+godot::Dictionary OpenXRFbHandTrackingAimExtensionWrapper::_get_requested_extensions() {
+	godot::Dictionary result;
+	for (auto ext : request_extensions) {
+		godot::String key = ext.first;
+		uint64_t value = reinterpret_cast<uint64_t>(ext.second);
+		result[key] = (godot::Variant)value;
+	}
+	return result;
+}
+
+void OpenXRFbHandTrackingAimExtensionWrapper::_on_state_ready() {
+	XRServer *xr_server = XRServer::get_singleton();
+
+	if (xr_server == nullptr) {
+		return;
+	}
+
+	trackers[Hand::HAND_LEFT].instantiate();
+	trackers[Hand::HAND_LEFT]->set_tracker_type(XRServer::TRACKER_CONTROLLER);
+	trackers[Hand::HAND_LEFT]->set_tracker_name("/user/fbhandaim/left");
+	trackers[Hand::HAND_LEFT]->set_tracker_desc("FB Aim tracker Left");
+	xr_server->add_tracker(trackers[Hand::HAND_LEFT]);
+
+	trackers[Hand::HAND_RIGHT].instantiate();
+	trackers[Hand::HAND_RIGHT]->set_tracker_type(XRServer::TRACKER_CONTROLLER);
+	trackers[Hand::HAND_RIGHT]->set_tracker_name("/user/fbhandaim/right");
+	trackers[Hand::HAND_RIGHT]->set_tracker_desc("FB Aim tracker Right");
+	xr_server->add_tracker(trackers[Hand::HAND_RIGHT]);
+}
+
+void OpenXRFbHandTrackingAimExtensionWrapper::_on_instance_destroyed() {
+	cleanup();
+}
+
+uint64_t OpenXRFbHandTrackingAimExtensionWrapper::_set_hand_joint_locations_and_get_next_pointer(int32_t p_hand_index, void *p_next_pointer) {
+	if (!fb_hand_tracking_aim_ext) {
+		return reinterpret_cast<uint64_t>(p_next_pointer);
+	}
+
+	aim_state[p_hand_index].type = XR_TYPE_HAND_TRACKING_AIM_STATE_FB;
+	aim_state[p_hand_index].next = p_next_pointer;
+	aim_state[p_hand_index].status = 0;
+
+	return reinterpret_cast<uint64_t>(&aim_state[p_hand_index]);
+}
+
+void OpenXRFbHandTrackingAimExtensionWrapper::_on_process() {
+	if (!is_enabled()) {
+		return;
+	}
+
+	for (int i = 0; i < Hand::HAND_MAX; i++) {
+		if (!trackers[i].is_valid()) {
+			continue;
+		}
+
+		XrPosef aim_pose = aim_state[i].aimPose;
+		XrQuaternionf aim_quat = aim_pose.orientation;
+		XrVector3f aim_position = aim_pose.position;
+		Quaternion quat = Quaternion(aim_quat.x, aim_quat.y, aim_quat.z, aim_quat.w);
+		Vector3 origin = Vector3(aim_position.x, aim_position.y, aim_position.z);
+
+		Transform3D transform = Transform3D(quat, origin);
+		Vector3 linear_velocity = Vector3(0.0, 0.0, 0.0);
+		Vector3 angular_velocity = Vector3(0.0, 0.0, 0.0);
+		XRPose::TrackingConfidence confidence = XRPose::TrackingConfidence::XR_TRACKING_CONFIDENCE_LOW;
+
+		if (!(aim_state[i].status & XR_HAND_TRACKING_AIM_VALID_BIT_FB)) {
+			confidence = XRPose::TrackingConfidence::XR_TRACKING_CONFIDENCE_NONE;
+		} else if (aim_state[i].status & XR_HAND_TRACKING_AIM_COMPUTED_BIT_FB) {
+			confidence = XRPose::TrackingConfidence::XR_TRACKING_CONFIDENCE_HIGH;
+		}
```

Comment:
Still more work to be done on updating this PR but I rebased and swapped over to positional trackers as suggested. Not sure if in this context there is a point to checking anything for tracking confidence, but for the moment I threw this in. I don't really understand what `XR_HAND_TRACKING_AIM_COMPUTED_BIT_FB` really means yet, from the spec on what the flag means:

> Aiming data is computed from additional sources beyond the hand data in the base structure

Reply:
I think that means that if full optical tracking isn't available (hand is not visible) that the current position is based on indirect measurements. I believe Quest for instance just added a way to determine hand position based on the cameras still seeing the arm.

So my gut feeling says that if `XR_HAND_TRACKING_AIM_COMPUTED_BIT_FB` is set, our confidence would be low, not high.

Could be wrong though and I don't think many users check beyond confidence not being `NONE`.

Reply:
I'd discussed with @dsnopek and he thought the same. At the moment I have it setting the confidence always to `LOW`, unless the flag `XR_HAND_TRACKING_AIM_VALID_BIT_FB` is not set, in which case I set confidence to `NONE`. So I'm not actually using the `XR_HAND_TRACKING_AIM_COMPUTED_BIT_FB` bit anywhere now.